synapse/push/httppusher.py

```diff
@@ -358,80 +379,115 @@ async def _build_notification_dict(
+            content["room_name"] = ctx["name"]
+
+        return (content, tweaks)
 
-        return d
+    def _build_notification_dict(
+        self,
+        content: JsonDict,
+        tweaks: Optional[JsonMapping],
+        default_payload: Optional[JsonMapping] = None,
+    ) -> JsonDict:
+        """Build a full notification from the content of it, as specified by the
+        Push Gateway spec. It wraps the content in a top level `notification`
+        element and put relevant device info in the `devices` section.
+
+        Args:
+            content: the content
+            tweaks: tweaks to add into the `devices` section
+            default_payload: default payload to add in `devices[0].data.default_payload`.
+            This will be merged (and override if some matching values already exist there)
+            with existing `default_payload`.
+
+        Returns:
+            a full notification that can be send to the registered push gateway.
+        """
+        content = content.copy()
+
+        data = self.data_minus_url.copy()
+        if default_payload:
+            if "default_payload" not in data:
+                data["default_payload"] = {}
+            data["default_payload"].update(default_payload)
+
+        device = {
+            "app_id": self.app_id,
+            "pushkey": self.pushkey,
+            "pushkey_ts": int(self.pushkey_ts / 1000),
+            "data": data,
+        }
+        if tweaks:
+            device["tweaks"] = tweaks
+
+        content["devices"] = [device]
+
+        return {"notification": content}
 
     async def dispatch_push(
-        self, event: EventBase, tweaks: Dict[str, bool], badge: int
+        self,
+        content: JsonDict,
+        tweaks: Optional[JsonMapping] = None,
+        default_payload: Optional[JsonMapping] = None,
     ) -> Union[bool, Iterable[str]]:
-        notification_dict = await self._build_notification_dict(event, tweaks, badge)
-        if not notification_dict:
-            return []
+        """Send a notification to the registered push gateway, with `content` being
+        the content of the `notification` top property specified in the spec.
+        If specified `devices` property will be overrided with device-specific
+        information for this pusher.
+
+        Args:
+            content: the content
+            tweaks: tweaks to add into the `devices` section
+            default_payload: default payload to add in `devices[0].data.default_payload`.
+            This will be merged (and override if some matching values already exist there)
+            with existing `default_payload`.
+
+        Returns:
+            False if an error occured when calling the push gateway, or an array of
+            rejected push keys otherwise. If this array is empty, the push fully
+            succeeded.
+        """
+        notif_dict = self._build_notification_dict(content, tweaks, default_payload)
```

Comment:
Would it be clearer if we inlined `_build_notification_dict` here?

Reply:
you are right, I am going to inline them. I kept it because it was there and didn't question it, but the `dispatch` methods are simple enough so I think it makes sense to inline.